Add test to check public APIs against master

driazati opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#60927**
* #61088

This implements/fixes #58617. If a PR adds a new bindings to the Python APIs, this will show up as a failing test, which they can fix by adding release notes to the PR body. The check errs on the side of caution in the face of flakiness, so if it can't reach S3 or doesn't have the env variables it expects it just quits.

Also this fixes an issue in the `torch._C` api bindings test where it wasn't actually testing the list defined in the file.

Differential Revision: [D29497017](https://our.internmc.facebook.com/intern/diff/D29497017)

driazati committed:
Add test to check public APIs against master

This implements/fixes #58617. If a PR adds a new bindings to the Python APIs, this will show up as a failing test, which they can fix by adding release notes to the PR body. The check errs on the side of caution in the face of flakiness, so if it can't reach S3 or doesn't have the env variables it expects it just quits.

Also this fixes an issue in the `torch._C` api bindings test where it wasn't actually testing the list defined in the file.

[ghstack-poisoned]

albanD commented on scripts/list_apis.py:
```diff
@@ -0,0 +1,195 @@
+                next_obj = getattr(obj, attr)
+                crawl_helper(next_obj, attr, path + [name], seen_objects, out, relevant_module)
+            except ModuleNotFoundError as e:
+                err_print(e)
```
Are these error useful? Or just triggering when we try to access attributes that we should not?

driazati replied:
mostly the errors come from objects / modules bound from C that Python can't deal with (e.g. missing attributes such as `__file__` on C modules) and some that explode if you try to access their attributes in any way (e.g. the TorchBind namespace). These are all generally private anyways so we don't really care

albanD commented on scripts/list_apis.py:
```diff
@@ -0,0 +1,195 @@
+                )
+
+    elif inspect.isclass(obj):
+        attrs = dir(obj)
```
You don't add the class itself here?

driazati replied:
it should be (like with modules), good catch

albanD commented on scripts/list_apis.py:
```diff
@@ -0,0 +1,195 @@
+
+def crawl_helper(
+    obj: Any,
+    name: str,
```
Is name always the same as path[-1] here ?

driazati replied:
`name` is the next attribute to look at in the `path` (which is just used for reporting back the fully qualified name), so on the next recursive call the previous `name` will be part of `path`

albanD commented on scripts/list_apis.py:
```diff
@@ -0,0 +1,195 @@
+CrawlResult = namedtuple("CrawlResult", ("public", "private", "errors", "module"))
+
+
+def err_print(*args: Any) -> None:
```
There are a lot of error printing happening here. Do we want to actually keep that? Or should we silence these? Or have a list of expected errors?

albanD reviewed (commented):
Thanks for working on this! Very interesting read!

driazati committed:
Update on "Add test to check public APIs against master"


This implements/fixes #58617. If a PR adds a new bindings to the Python APIs, this will show up as a failing test, which they can fix by adding release notes to the PR body. The check errs on the side of caution in the face of flakiness, so if it can't reach S3 or doesn't have the env variables it expects it just quits.

Also this fixes an issue in the `torch._C` api bindings test where it wasn't actually testing the list defined in the file.

[ghstack-poisoned]

driazati commented:
@driazati has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29497017).

driazati committed:
Update on "Add test to check public APIs against master"


This implements/fixes #58617. If a PR adds a new bindings to the Python APIs, this will show up as a failing test, which they can fix by adding release notes to the PR body. The check errs on the side of caution in the face of flakiness, so if it can't reach S3 or doesn't have the env variables it expects it just quits.

Also this fixes an issue in the `torch._C` api bindings test where it wasn't actually testing the list defined in the file.

Differential Revision: [D29497017](https://our.internmc.facebook.com/intern/diff/D29497017)

[ghstack-poisoned]

driazati committed:
Update on "Add test to check public APIs against master"


This implements/fixes #58617. If a PR adds a new bindings to the Python APIs, this will show up as a failing test, which they can fix by adding release notes to the PR body. The check errs on the side of caution in the face of flakiness, so if it can't reach S3 or doesn't have the env variables it expects it just quits.

Also this fixes an issue in the `torch._C` api bindings test where it wasn't actually testing the list defined in the file.

Differential Revision: [D29497017](https://our.internmc.facebook.com/intern/diff/D29497017)

[ghstack-poisoned]

driazati committed:
Update on "Add test to check public APIs against master"


This implements/fixes #58617. If a PR adds a new bindings to the Python APIs, this will show up as a failing test, which they can fix by adding release notes to the PR body. The check errs on the side of caution in the face of flakiness, so if it can't reach S3 or doesn't have the env variables it expects it just quits.

Also this fixes an issue in the `torch._C` api bindings test where it wasn't actually testing the list defined in the file.

Differential Revision: [D29497017](https://our.internmc.facebook.com/intern/diff/D29497017)

[ghstack-poisoned]

driazati commented:
@driazati has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29497017).

driazati committed:
Update on "Add test to check public APIs against master"


This implements/fixes #58617. If a PR adds a new bindings to the Python APIs, this will show up as a failing test, which they can fix by adding release notes to the PR body. The check errs on the side of caution in the face of flakiness, so if it can't reach S3 or doesn't have the env variables it expects it just quits.

Also this fixes an issue in the `torch._C` api bindings test where it wasn't actually testing the list defined in the file.

Differential Revision: [D29497017](https://our.internmc.facebook.com/intern/diff/D29497017)

[ghstack-poisoned]

driazati committed:
Update on "Add test to check public APIs against master"


This implements/fixes #58617. If a PR adds a new bindings to the Python APIs, this will show up as a failing test, which they can fix by adding release notes to the PR body. The check errs on the side of caution in the face of flakiness, so if it can't reach S3 or doesn't have the env variables it expects it just quits.

Also this fixes an issue in the `torch._C` api bindings test where it wasn't actually testing the list defined in the file.

Differential Revision: [D29497017](https://our.internmc.facebook.com/intern/diff/D29497017)

[ghstack-poisoned]

walterddr reviewed (commented):
I might not be the best to review. but did we consider reusing some logic from target_determinator, which uses modulefinder to crawl the entire codebase backwards from test files. I haven't tried, but i think it could be adjust to crawl from `torch`?

samestep commented on .github/templates/linux_ci_workflow.yml.j2:
```diff
@@ -493,3 +495,80 @@ jobs:
+      - name: Checkout PyTorch
+        uses: actions/checkout@v2
+        with:
+          fetch-depth: 0 # deep clone, to allow sharding to use git rev-list
```
this job doesn't do anything with sharding, right? could you either remove this line or modify this comment to clarify why it's necessary?

samestep commented on .github/templates/linux_ci_workflow.yml.j2:
```diff
@@ -493,3 +495,80 @@ jobs:
+      - name: Archive APIs
+        run: |
+          zip apis.zip apis.log
+      - uses: driazati/upload-artifact-s3@50adbe4ef0b6d9221df25c18c5fc528dfcb7c3f8
```
it looks like there are three different versions of `upload-artifact-s3` that we use:

- `driazati/upload-artifact-s3@50adbe4ef0b6d9221df25c18c5fc528dfcb7c3f8` (this one), in one place
- `driazati/upload-artifact-s3@21c31d0a7bcb056ca50bd6ce197ba6507c26a1be`, in [one place](https://github.com/pytorch/pytorch/blob/4a544df00dee565e6ccb73e971ff734346b16474/.github/templates/linux_ci_workflow.yml.j2#L484)
- `seemethere/upload-artifact-s3@9d7ceb0ab39c2c88d93ef7792b27425b27d59162`, in [two](https://github.com/pytorch/pytorch/blob/4a544df00dee565e6ccb73e971ff734346b16474/.github/templates/linux_ci_workflow.yml.j2#L188) [places](https://github.com/pytorch/pytorch/blob/4a544df00dee565e6ccb73e971ff734346b16474/.github/templates/windows_ci_workflow.yml.j2#L85)

If it is necessary to use all these different versions, could you add code comment(s) explaining why? And if not, would it be a good idea to consolidate to use a single version?

driazati replied:
We should definitely consolidate: #61384

samestep commented:
why is `list_apis.py` in `scripts/`? that causes it to not be checked by either Flake8 or `mypy`

samestep commented on scripts/list_apis.py:
```diff
@@ -0,0 +1,277 @@
+    )
+    parser.add_argument("--module", help="module to crawl", required=True)
+
+    def add_flag(name: str, default: bool, help: str) -> None:
+        parser.add_argument(f"--{name}", dest=name, help=help, action="store_true")
+        parser.add_argument(f"--no-{name}", dest=name, help=help, action="store_false")
+        parser.set_defaults(**{name: default})
```
I don't quite understand this: in `main` above, you use this for `public`, `private`, and `errors`, but in `main` above, you implicitly coerce all three of those to booleans before using them, so the `--no-` variant seems to not do anything; could you clarify why you add it?

samestep replied:
oh never mind, sorry; I didn't see that you set the `default` for these when you call `add_flag` below

samestep commented on scripts/list_apis.py:
```diff
@@ -0,0 +1,277 @@
+
+    def should_skip(self, obj: Any, path: List[Item]) -> bool:
+        """
+        Return whether 'obj' is relevant to 'relevant_module'. If 'obj' is not a
+        module, then get its module from the __module__ attribute. Then check if
+        the module is part of the relevant_module by comparing their __file__ paths.
```
what is `relevant_module`?

driazati replied:
forgot to update the comment after OOP-ing everything, it's the main module that's being tracked (so this should be `relevant_module` -> `self.module`)

samestep commented on scripts/list_apis.py:
```diff
@@ -0,0 +1,277 @@
+            curr_len = len(path) + 1
+            if curr_len >= prev_len:
+                # It's a longer path and this already has an entry, so don't
+                # both with it any further
```
typo I think
```suggestion
                # bother with it any further
```

samestep commented on scripts/list_apis.py:
```diff
@@ -0,0 +1,277 @@
+
+
+        if module == self.module:
+            # If the object is a direct child of the concered module, then don't
```
typo I think
```suggestion
            # If the object is a direct child of the concerned module, then don't
```

samestep commented on scripts/list_apis.py:
```diff
@@ -0,0 +1,277 @@
+
+    def get_submodules(self, module: ModuleType) -> List[ModuleType]:
+        """
+        Some modules aren't imported directly into the parent (i.e. torch.fx). So
```
```suggestion
        Some modules aren't imported directly into the parent (e.g. torch.fx). So
```

samestep commented on scripts/list_apis.py:
```diff
@@ -0,0 +1,277 @@
+            if inspect.isclass(obj):
+                # We don't want to re-add class attributes (these don't have an
+                # associated id() to de-duplicate them with, so we have to do it
+                # manually
```
```suggestion
                # manually)
```

samestep commented on test/test_public_bindings.py:
```diff
@@ -1,10 +1,81 @@
+
+                ### Release Notes
+                * the new feature""")
+        self.assertTrue(difference != set() and "### Release Notes" not in PR_BODY, msg)
```
I'm a bit confused about this; shouldn't it instead look like the following?
```suggestion
        self.assertTrue(difference == set() or "### Release Notes" in PR_BODY, msg)
```

driazati commented:
> modulefinder

`modulefinder` kind of does the same thing (though it has a bug where it crashes on the `mpl_toolkits` module), but it misses modules unreachable from the source like `torch.fx`

```python

driazati committed:
Update on "Add test to check public APIs against master"


This implements/fixes #58617. If a PR adds a new bindings to the Python APIs, this will show up as a failing test, which they can fix by adding release notes to the PR body. The check errs on the side of caution in the face of flakiness, so if it can't reach S3 or doesn't have the env variables it expects it just quits.

Also this fixes an issue in the `torch._C` api bindings test where it wasn't actually testing the list defined in the file.

Differential Revision: [D29497017](https://our.internmc.facebook.com/intern/diff/D29497017)

[ghstack-poisoned]

driazati commented:
closing since this seems like a recipe for flakiness
